### PR TITLE
Remove registry storage testdriver

### LINKED ISSUE #2151: Remove `registry/storage/driver/testdriver`
The test driver should be removed and replaced with the `inmemory` driver. The `inmemory` driver should be updated to fully comply with the storagedriver interface to ensure that any written data is not made available until commit has been called.

### PR DESCRIPTION
**Pull request by glefloch**
This PR aims to close #2151 by removing the storage testdriver and using the inmemory driver. 
This also update the inmemory storage driver to make sure that any written data is not available before commit.

Signed-off-by: glefloch <glfloch@gmail.com>

### CODE REVIEW

**Comment by milosgajdos** [comment]
@glefloch can you please rebase?

**Comment by milosgajdos** [comment]
Ping @glefloch

File: registry/handlers/api_test.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 	"strings"
 	"testing"
 
+<<<<<<< HEAD
```
Comment (milosgajdos) [inline registry/handlers/api_test.go:23]: Hm, yeah, this won't work 😅

**Comment by milosgajdos** [comment]
I think you might need to run `go mod tidy` and `go mod vendor` 🤔

**Comment by milosgajdos** [comment]
@glefloch are you still interested in getting this PR done? There are some merge conflicts that need fixing

**Comment by glefloch** [comment]
I need to set up my computer but yes, still interested into getting this merged!

File: registry/handlers/api_test.go [hunk 2]
```diff
@@ -31,8 +31,8 @@ import (
 	v2 "github.com/distribution/distribution/v3/registry/api/v2"
 	storagedriver "github.com/distribution/distribution/v3/registry/storage/driver"
 	"github.com/distribution/distribution/v3/registry/storage/driver/factory"
-	_ "github.com/distribution/distribution/v3/registry/storage/driver/testdriver"
 	"github.com/distribution/distribution/v3/testutil"
+	_ "github.com/docker/distribution/registry/storage/driver/inmemory"
```
Comment (thaJeztah) [inline registry/handlers/api_test.go:35]: This should use the new location; using the old name causes the code from this repository (but the v2 release) to be vendored;


```go
	_ "github.com/distribution/distribution/v3/registry/storage/driver/inmemory"
```

File: registry/handlers/app_test.go [hunk 3]
```diff
@@ -16,15 +16,15 @@ import (
 	_ "github.com/distribution/distribution/v3/registry/auth/silly"
 	"github.com/distribution/distribution/v3/registry/storage"
 	memorycache "github.com/distribution/distribution/v3/registry/storage/cache/memory"
-	"github.com/distribution/distribution/v3/registry/storage/driver/testdriver"
+	"github.com/docker/distribution/registry/storage/driver/inmemory"
```
Comment (thaJeztah) [inline registry/handlers/app_test.go:19]: Same here;

```suggestion
	"github.com/distribution/distribution/v3/registry/storage/driver/inmemory"
```

File: registry/storage/blob_test.go [hunk 4]
```diff
@@ -14,8 +14,8 @@ import (
 	"github.com/distribution/distribution/v3"
 	"github.com/distribution/distribution/v3/reference"
 	"github.com/distribution/distribution/v3/registry/storage/cache/memory"
-	"github.com/distribution/distribution/v3/registry/storage/driver/testdriver"
 	"github.com/distribution/distribution/v3/testutil"
+	"github.com/docker/distribution/registry/storage/driver/inmemory"
```
Comment (thaJeztah) [inline registry/storage/blob_test.go:18]: And here

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: go.sum [hunk 5]
```diff
@@ -59,6 +59,8 @@ github.com/beorn7/perks v1.0.1 h1:VlbKKnNfV8bJzeqoa4cOKqO6bYr3WgKZxO8Z16+hsOM=
 github.com/beorn7/perks v1.0.1/go.mod h1:G2ZrVWU2WbWT9wwq4/hrbKbnv/1ERSJQ0ibhJ6rlkpw=
 github.com/bitly/go-simplejson v0.5.0 h1:6IH+V8/tVMab511d5bn4M7EwGXZf9Hj6i2xSwkNEM+Y=
 github.com/bitly/go-simplejson v0.5.0/go.mod h1:cXHtHw4XUPsvGaxgjIAn8PhEWG9NfngEKAMDJEczWVA=
+github.com/bmizerany/assert v0.0.0-20160611221934-b7ed37b82869 h1:DDGfHa7BWjL4YnC6+E63dPcxHo2sUxDIu8g3QgEJdRY=
+github.com/bmizerany/assert v0.0.0-20160611221934-b7ed37b82869/go.mod h1:Ekp36dRnpXw/yCqJaO+ZrUyxD+3VXMFFr56k5XYrpB4=
```
Comment (thaJeztah) [inline go.sum:63]: This is probably a left-over from your previous iteration (which was importing the old `github.com/docker/distribution`)

File: vendor/github.com/docker/distribution/LICENSE [hunk 6]
```diff
@@ -0,0 +1,202 @@
+Apache License
```
Comment (thaJeztah) [inline vendor/github.com/docker/distribution/LICENSE:1]: Same for all these files in `vendor`

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks!

Looks like there's still vendor changes related to your previous iteration; can you run;

```bash
go mod tidy
go mod vendor
```

And amend your commit with those changes?

File: registry/storage/blob_test.go
Comment (thaJeztah) [inline registry/storage/blob_test.go:1]: Looks like in this file you only updated the import, but forgot to replace references to the test-driver; for example, see line 26 (and some others);

https://github.com/distribution/distribution/blob/788598e23f50db1b2f89e780683a7a6a0eadf3d1/registry/storage/blob_test.go#L26-L26

**Comment by glefloch** [comment]
It looks better thanks to `go mod` commands

**Review by thaJeztah — APPROVED** [review]
LGTM

ping @milosgajdos @corhere ptal

**Comment by corhere** [comment]
@squizzi PTAL

**Pull request merged**